src/theory/sets/inference_manager.cpp

```diff
@@ -96,26 +81,30 @@ bool InferenceManager::assertFactRec(Node fact, Node exp, int inferType)
   }
   bool polarity = fact.getKind() != NOT;
   TNode atom = polarity ? fact : fact[0];
+  if (d_state.isEntailed(atom, polarity))
```

Comment:
How about moving this entailment checking to the top of the function. 
There is another entailment checking in line 40, which can be refactored with this one in line 84

Reply:
The issue here is performance.  Note there are two different entailment checks here: either the entire formula `fact` is ready to be sent out as a formula, in which case we run an entailment check (once) on the full formula.  

Otherwise, we are traversing to each position in the formula (e.g. based on AND), and checking entailment when we get to a literal.  If we were to check at the beginning of the method, we would potentially be traversing on the order of `n^2` terms (checking entailment for each subformula) where `n` is the size of fact instead of just `n` (checking each leaf).